src/sqlparser/src/parser.rs

```diff
@@ -583,6 +583,21 @@ impl Parser {
         }
     }
 
+    fn parse_param(&mut self, param: String) -> Result<Expr, ParserError> {
+        let data_type = if let Token::DoubleColon = self.peek_token() {
+            self.next_token();
+            Some(self.parse_data_type()?)
+        } else {
+            None
+        };
```

Comment:
Why do we include the cast as part of parameter? Is it possible to be `CAST($1 AS int)` rather than `$1::int`?

One idea is to just parse the outer part as a normal cast, which includes both `CAST(AS)` and `::`. Then during binder type inference, just treat `$1` as `unknown` (same as `null` and `'12'`).

^ This can also be a refactor after we have build the whole code path of this paramater-with-type first.

Reply:
I think this is the right way.🤣 Let's do it now.